### PR TITLE
AAP-4359 Maintain Hub API during AAP upgrade + basic upgrade section

### PR DESCRIPTION
**Pull request by TarikFD**
Jira: https://issues.redhat.com/browse/AAP-4359

Request was to document feature in the installer for users to import an existing Hub API token or generate a new one using the AAP installer during an AAP upgrade. In doing so, I'm adding a section to outline a "basic upgrade" of AAP to guide users through editing the inventory file and running the setup script. Working with product team to ensure this new section makes sense.

Preview (VPN required): http://file.rdu.redhat.com/kevchin/aap-upgrade6/tmp/en-US/html-single/

### CODE REVIEW

**Comment by TarikFD** [comment]
@oraNod Please give the organization a review and let me know if the information flow makes sense. The changes I've made:

Added chapter 1 that guides users through different upgrade paths. Users upgrading to 2.2 need to do different things depending on their current version, so chapter 1 links to the appropriate sections. 

Also added chapter 2 that covers the "basic upgrade" path, as it wasn't documented before how a user would perform an upgrade from 2.1 to 2.2 .

File: downstream/modules/platform/con-aap-upgrade-from-20.adoc [hunk 1]
```diff
@@ -6,9 +6,10 @@
...
 
-* <<migrate-isolated-execution-nodes,Migrate isolated nodes to execution nodes>>
-**  Migrate data from isolated legacy nodes to execution nodes by performing a side-by-side upgrade to be compatible with the latest automation mesh features
 * <<upgrading-to-ees,Migrate virtual environments to automation execution environments>>
 ** Duplicate your previous custom virtual environments into new automation execution environments using the `awx-manage` command.
+* <<migrate-isolated-execution-nodes,Migrate isolated nodes to execution nodes>>
+**  Migrate data from isolated legacy nodes to execution nodes by performing a side-by-side upgrade to be compatible with the latest automation mesh features
+** Import or generate a new {HubName} API token
 * <<content-migration,Migrate Ansible content>>
 ** Reconfigure your Ansible content to include Fully Qualified Collection Names (FQCN) to be compatible with `ansible-core` 2.13.
```
Comment (oraNod) [inline downstream/modules/platform/con-aap-upgrade-from-20.adoc:15]: I'll do some edits and push a commit but I have a couple nits here. The main one is that this module feels more like a procedure than a concept, especially with the gerund title.

**Review by oraNod — COMMENTED** [review]
@TarikFD Hey, thanks for adding me as a reviewer on this one. I'm going to submit a couple of comments now but am still working on a few things. I need to step away for a bit but will finish this up by EOD. I'll also send a commit with a few suggestions.

File: downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc [hunk 2]
```diff
@@ -8,7 +8,7 @@ ifdef::context[:parent-context: {context}]
 
 [role="_abstract"]
 
-The process of upgrading to {PlatformName} {PlatformVers} will differ depending on the version of {PlatformNameShort} you are upgrading from. Upgrade from an older version of {PlatformNameShort} (version 2.0 or earlier) by migrating legacy Ansible content to be compatible with the latest features of {PlatformNameShort} {PlatformVers}.
+The process of upgrading to {PlatformName} {PlatformVers} will differ depending on the version of {PlatformNameShort} you are upgrading from. Upgrade from an older version of {PlatformNameShort} (version 2.0 or earlier) by migrating legacy Ansible content to be compatible with the latest features of {PlatformNameShort} {PlatformVers}. Otherwise, upgrade to {PlatformName} {PlatformVers} by performing a basic upgrade.
 
 include::platform/con-aap-upgrade-from-20.adoc[leveloffset=+1]
 include::platform/con-aap-upgrade-from-21.adoc[leveloffset=+1]
```
[the remark this answers is not in the document]
Comment (oraNod) [reply downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc:16]: It might also be worth flipping the order and putting 2.1 upgrades before legacy upgrades.

File: downstream/modules/platform/con-aap-upgrade-from-21.adoc [hunk 3]
```diff
@@ -1,11 +1,15 @@
 
 
-[id="aap-upgrade-from-21_{context}"]
+[id="aap-upgrades_{context}"]
 
-= Upgrading from {PlatformNameShort} 2.1
+= {PlatformNameShort} upgrades
```
Comment (oraNod) [inline downstream/modules/platform/con-aap-upgrade-from-21.adoc:5]: Looking at the generated HTML there are other titles in this assembly that I think we should align to if you want to go ahead with this change. I didn't edit them in my commit because I didn't want to blow up the scope of your PR.

File: downstream/modules/platform/con-aap-upgrade-from-21.adoc [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
 
-Upgrade from {PlatformNameShort} version 2.1 to version {PlatformVers} by following all the procedures in these sections:
+Upgrading to version {PlatformVers} from {PlatformNameShort} 2.1 or later involves downloading the installation package and then performing the following steps:
 
+* Set up your inventory to match your installation environment.
+* Run the {PlatformVers} installation program over your current {PlatformNameShort} installation.
+* Import or generate a new {HubName} API token.
+
+[role="_additional-resources"]
+.Additional resources
```
Comment (oraNod) [inline downstream/modules/platform/con-aap-upgrade-from-21.adoc:14]: I'm a big, big fan of the additional resources section. Using inline links adds noise to the content and risks getting users off track.

The additional resources section means that readers are much more likely to go through the full text in the topic - and not miss anything essential. It also trains the reader to look for links in the same place. After a bit of navigation, readers generally know where to go to find their link.

File: downstream/modules/platform/ini/clustered-nodes.ini [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+[all:vars]
+admin_password='password'
+
+pg_host=''
+pg_port=''
+
+pg_database='awx'
+pg_username='awx'
+pg_password='password'
```
Comment (oraNod) [inline downstream/modules/platform/ini/clustered-nodes.ini:14]: Maybe this should be `<password>` to indicate a replaceable value. You could also go with `<changeme>` to make it more obvious that it's a value that the user needs to input and not a literal value.

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 6]
```diff
@@ -33,7 +33,12 @@ The setup script pauses and indicates that a "pre-2.x" inventory file was detect
 NOTE: By running the setup script, the Installer modified a few fields from your original inventory file, such as renaming [tower] to [automationcontroller].
 . Modify the newly generated `inventory.new.ini` file to configure your automation mesh by assigning relevant variables, nodes, and relevant node-to-node peer connections:
 +
-NOTE: The design of your automation mesh topology depends on the automation needs of your environment. The example below offers one possible scenario for automation mesh design.
+[NOTE]
+====
+The design of your {AutomationMesh} topology depends on the automation needs of your environment.
+It is beyond the scope of this document to provide designs for all possible scenarios.
+The following is one example {AutomationMesh} design.
```
Comment (oraNod) [inline downstream/modules/platform/proc-upgrade-installer.adoc:40]: For admonitions that include more than one line of text I think the convention is to use the format:

```
[ADMONITION_TYPE]
====
Lorem ipsum
====
```

Comment (oraNod) [reply downstream/modules/platform/proc-upgrade-installer.adoc:40]: I'm also a bit of a stickler for keeping each sentence on a separate line. Asciidoc automatically forms paragraphs from consecutive lines of text. Putting each sentence on a separate line makes it easier to review files.

File: downstream/modules/platform/ini/clustered-nodes.ini (hunk 5 above)
Comment (TarikFD) [reply downstream/modules/platform/ini/clustered-nodes.ini:14]: Noted about the replaceable value. Is there a reason why we'd be calling a seperate .ini file for the code block rather than including the code block in the main doc itself?

File: downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+[id="aap-upgrade-to-latest"]
+= {PlatformName} {PlatformVers} Upgrades
+
+:context: aap-upgrade-to-latest
+
+[role="_abstract"]
+
+Upgrade to {PlatformName} {PlatformVers} by setting up your inventory and running the installation script.
+Ansible then upgrades your deployment to {PlatformVers}.
+If you plan to upgrade from {PlatformNameShort} 2.0 or earlier, you need to migrate Ansible content for compatibility with {PlatformVers}.
```
Comment (ianf77) [inline downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc:13]: , you must migrate Ansible content

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 8]
```diff
@@ -7,18 +7,21 @@ To upgrade your instance of Ansible Tower to {PlatformNameShort} {PlatformVers},
...
 ----
-$ cd ansible-automation-platform-setup-<latest-version>/
+$ cd ansible-automation-platform-setup-__<latest_version__>/
 ----
 . Copy the `inventory` file from your original instance into the directory of the latest installer:
 +
+[subs="+quotes"]
 ----
-$ cp ansible-tower-setup-3.8.x.x/inventory ansible-automation-platform-setup-<latest-version>
+$ cp ansible-tower-setup-3.8.x.x/inventory ansible-automation-platform-setup-__<latest_version__>
```
Comment (oraNod) [inline downstream/modules/platform/proc-upgrade-installer.adoc:24]:
```suggestion
$ cp ansible-tower-setup-3.8.x.x/inventory ansible-automation-platform-setup-__<latest_version>__
```

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 9]
```diff
@@ -7,18 +7,21 @@ To upgrade your instance of Ansible Tower to {PlatformNameShort} {PlatformVers},
...
 ----
-$ tar xvzf ansible-automation-platform-setup-<latest-version>.tar.gz
+$ tar xvzf ansible-automation-platform-setup-__<latest_version__>.tar.gz
 ----
 . Navigate into your {PlatformNameShort} installation directory:
 +
+[subs="+quotes"]
 ----
-$ cd ansible-automation-platform-setup-<latest-version>/
+$ cd ansible-automation-platform-setup-__<latest_version__>/
```
Comment (oraNod) [inline downstream/modules/platform/proc-upgrade-installer.adoc:18]:
```suggestion
$ cd ansible-automation-platform-setup-__<latest_version>__/
```

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 10]
```diff
@@ -64,6 +72,21 @@ node_type=execution
 <1> Specifies a control node that runs project and inventory updates and system jobs, but not regular jobs. Execution capabilities are disabled on these nodes.
 <2> Specifies peer relationships for node-to-node connections in the `[execution_nodes]` group.
 <3> Specifies hop nodes that route traffic to other execution nodes. Hop nodes cannot execute automation.
+. Import or generate a {HubName} API token.
++
+Import an existing API token with the `automationhub_api_token` flag:
```
Comment (oraNod) [inline downstream/modules/platform/proc-upgrade-installer.adoc:77]:
```suggestion
* Import an existing API token with the `automationhub_api_token` flag:
```

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 11]
```diff
@@ -64,6 +72,21 @@ node_type=execution
...
++
+[source,options="nowrap",subs=attributes+]
+----
+automationhub_api_token=<api_token>
+----
++
+Generate a new API token, and invalidate any existing tokens, with the `generate_automationhub_token` flag:
```
Comment (oraNod) [inline downstream/modules/platform/proc-upgrade-installer.adoc:84]:
```suggestion
* Generate a new API token, and invalidate any existing tokens, with the `generate_automationhub_token` flag:
```

File: downstream/assemblies/platform/assembly-migrate-isolated-execution-nodes.adoc [hunk 12]
```diff
@@ -1,6 +1,6 @@
 ifdef::context[:parent-context: {context}]
 
-[id="assembly-migrate-isolated-execution-nodes"]
+[id="migrate-isolated-execution-nodes"]
 = Upgrading isolated nodes to execution nodes
```
Comment (oraNod) [inline downstream/assemblies/platform/assembly-migrate-isolated-execution-nodes.adoc:4]: Title does not match the ID or filename.

File: downstream/assemblies/platform/assembly-aap-upgrading-platform.adoc [hunk 13]
```diff
@@ -1,26 +1,17 @@
...
-To upgrade your {PlatformName} to the later version, download the desired version of the {PlatformNameShort} installer. Configure it with the same parameters you have used in previous installations, configure any new parameters, then run the installer.
-
-In some cases, you may need to configure other settings if you plan on upgrading other components of your {PlatformNameShort}, such as {ControllerName}. The following sections describes these steps in further detail.
+To perform a upgrade of your {PlatformName} to a later version, download the desired version of the {PlatformNameShort} installer. Configure the `inventory` file with the same parameters you have used in previous installations and add new parameters to enable any new features, then run the installer.
 
-include::platform/con-upgrade-planning.adoc[leveloffset=+1]
+include::platform/con-aap-upgrade-planning.adoc[leveloffset=+1]
 
-Once you have completed the necessary changes during upgrade planning, proceed by downloading the desired version of the {PlatformNameShort} installer.
+Once you have completed the necessary upgrade planning, proceed by downloading the desired version of the {PlatformNameShort} installer.
```
Comment (oraNod) [inline downstream/assemblies/platform/assembly-aap-upgrading-platform.adoc:14]:
```suggestion
After you have completed the necessary upgrade planning, proceed by downloading the desired version of the {PlatformNameShort} installer.
```

File: downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc [hunk 14]
```diff
@@ -0,0 +1,20 @@
+
+ifdef::context[:parent-context: {context}]
+
+[id="aap-upgrade-to-latest"]
+= {PlatformName} {PlatformVers} Upgrades
```
Comment (oraNod) [inline downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc:5]:
```suggestion
= {PlatformName} {PlatformVers} upgrades
```

Comment (oraNod) [reply downstream/assemblies/platform/assembly-aap-upgrade-to-latest.adoc:5]: Title doesn't match the ID or filename. I would just change the ID and filename to "aap-upgrades".

File: downstream/modules/platform/ini/clustered-nodes.ini (hunk 5 above)
Comment (oraNod) [reply downstream/modules/platform/ini/clustered-nodes.ini]: There are a few reasons to externalize content to native file types. Doing this makes it easy to lint and validate automatically. For instance if you externalize YAML content to `yaml/myfile.yaml` you can lint it in your IDE to make sure the syntax is valid. You can also lint as part of the PR process. We could set things up to validate YAML, JSON, et al whenever someone pushes a commit.

Another good reason is reuse. You can include the same file in multiple modules. I'm hoping we'll get to the point where a lot of the upstream docs are externalizing files too and then we'll be able to include them in our asciidoc, if writers get into the habit of externalizing.

Another really compelling reason for externalizing files, is that engineers will be more likely to spot issues a piece of code in the correct format rather than embedded in asciidoc. Engineers, QE, and support teams are also more likely to contribute and help maintain externalized content and take responsibility for it. Writers are the SMEs for asciidoc but not necessarily for things like YAML or whatever else.

File: downstream/modules/platform/con-aap-upgrades.adoc [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+[id="aap-upgrades_{context}"]
+
+= {PlatformNameShort} upgrades
+
+Upgrading to version {PlatformVers} from {PlatformNameShort} 2.1 or later involves downloading the installation package and then performing the following steps:
+
+* Set up your inventory to match your installation environment.
+* Run the {PlatformVers} installation program over your current {PlatformNameShort} installation.
+* Import or generate a new {HubName} API token.
```
Comment (simaishi) [inline downstream/modules/platform/con-aap-upgrades.adoc:11]: This is not 'required' when upgrading from 2.1. Although you 'can' still import or generate api token for 2.1 -> 2.2 upgrade, I feel mentioning will cause confusion as most customers probably won't do it. Can you reword (or move section or whatever...) so it's clear this is required for upgrade from 1.2/2.0 only?

File: downstream/modules/platform/con-aap-upgrade-planning.adoc [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+== {ControllerNameStart}
+
+* Even if you have a valid license from a previous version, you must provide your credentials or a subscriptions manifest upon upgrading to the latest version of {ControllerName}.
+* If you need to upgrade Red Hat Enterprise Linux and {ControllerName}, you must first backup and restore of your {ControllerName} data.
+* Clustered upgrades require special attention to instance and instance groups before upgrading.
+
+[role="_additional-resources"]
+.Additional resources
+* link:https://docs.ansible.com/automation-controller/latest/html/userguide/import_license.html[Import a subscription]
+* link:https://docs.ansible.com/automation-controller/latest/html/administration/backup_restore.html#ag-backup-restore[Back up and restoring Controller]
```
Comment (ariordan-redhat) [inline downstream/modules/platform/con-aap-upgrade-planning.adoc:20]:
```suggestion
* link:https://docs.ansible.com/automation-controller/latest/html/administration/backup_restore.html#ag-backup-restore[Backing up and restoring Controller]
```

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 17]
```diff
@@ -64,6 +72,21 @@ node_type=execution
...
+[source,options="nowrap",subs=attributes+]
+----
+automationhub_api_token=<api_token>
+----
++
+* Generate a new API token, and invalidate any existing tokens, with the `generate_automationhub_token` flag:
++
+[source,options="nowrap",subs=attributes+]
+----
+generate_automationhub_token=
```
Comment (ariordan-redhat) [inline downstream/modules/platform/proc-upgrade-installer.adoc:88]:
```suggestion
generate_automationhub_token=true
```
From "- set 'generate_automationhub_token' to True to generate a new token" in the ticket. Don't know whether it should be `true`, `"true"` or `True`.

**Review by ariordan-redhat — CHANGES_REQUESTED** [review]
Noticed a couple of typos. Other comments are optional suggestions

**Review by dcdacosta — APPROVED** [review]
Other than correcting the items Aine commented on, LGTM.

File: downstream/assemblies/platform/assembly-aap-upgrading-platform.adoc [hunk 18]
```diff
@@ -1,26 +1,17 @@
...
-To upgrade your {PlatformName} to the later version, download the desired version of the {PlatformNameShort} installer. Configure it with the same parameters you have used in previous installations, configure any new parameters, then run the installer.
-
-In some cases, you may need to configure other settings if you plan on upgrading other components of your {PlatformNameShort}, such as {ControllerName}. The following sections describes these steps in further detail.
+To perform a upgrade of your {PlatformName} to a later version, download the desired version of the {PlatformNameShort} installer. Configure the `inventory` file with the same parameters you have used in previous installations and add new parameters to enable any new features, then run the installer.
 
-include::platform/con-upgrade-planning.adoc[leveloffset=+1]
+include::platform/con-aap-upgrade-planning.adoc[leveloffset=+1]
 
-Once you have completed the necessary changes during upgrade planning, proceed by downloading the desired version of the {PlatformNameShort} installer.
+After you have completed the necessary upgrade planning, proceed by downloading the desired version of the {PlatformNameShort} installer.
```
Comment (lburcher7) [inline downstream/assemblies/platform/assembly-aap-upgrading-platform.adoc:14]:
```suggestion
After you have completed the necessary upgrade planning, download the desired version of the {PlatformNameShort} installer.
```

**Review by lburcher7 — COMMENTED** [review]
Test review.

**Review by ariordan-redhat — APPROVED** [review]
LGTM

File: downstream/assemblies/platform/assembly-aap-upgrading-platform.adoc [hunk 19]
```diff
@@ -1,26 +1,17 @@
...
-To upgrade your {PlatformName} to the later version, download the desired version of the {PlatformNameShort} installer. Configure it with the same parameters you have used in previous installations, configure any new parameters, then run the installer.
-
-In some cases, you may need to configure other settings if you plan on upgrading other components of your {PlatformNameShort}, such as {ControllerName}. The following sections describes these steps in further detail.
+To perform an upgrade of your {PlatformName} to a later version, download the desired version of the {PlatformNameShort} installer. Configure the `inventory` file with the same parameters you have used in previous installations and add new parameters to enable any new features, then run the installer.
 
-include::platform/con-upgrade-planning.adoc[leveloffset=+1]
+include::platform/con-aap-upgrade-planning.adoc[leveloffset=+1]
 
-Once you have completed the necessary changes during upgrade planning, proceed by downloading the desired version of the {PlatformNameShort} installer.
+After you have completed the necessary upgrade planning, download the desired version of the {PlatformNameShort} installer.
```
Comment (oraNod) [inline downstream/assemblies/platform/assembly-aap-upgrading-platform.adoc:14]: I'm not sure about putting sentences between include directives in an assembly like this. Perhaps you could combine this with the first sentence of the abstract.

To upgrade your {PlatformName}, start by reviewing planning information to ensure a successful upgrade.
You can then download the desired version of the {PlatformNameShort} installer, configure the `inventory` file in the installation bundle to reflect your environment, and then run the installer.

File: downstream/assemblies/platform/assembly-migrate-isolated-execution-nodes.adoc [hunk 20]
```diff
@@ -1,7 +1,7 @@
 ifdef::context[:parent-context: {context}]
 
-[id="assembly-migrate-isolated-execution-nodes"]
-= Upgrading isolated nodes to execution nodes
+[id="migrate-isolated-execution-nodes"]
+= Migrating isolated nodes to execution nodes
 
 [role="_abstract"]
 Upgrading from version 1.x to the latest version of the {PlatformName} requires platform administrations to migrate data from isolated legacy nodes to execution nodes. This migration is necessary to deploy the automation mesh.
```
Comment (oraNod) [inline downstream/assemblies/platform/assembly-migrate-isolated-execution-nodes.adoc:7]:
```suggestion
Upgrading from version 1.x to the latest version of the {PlatformName} requires platform administrators to migrate data from isolated legacy nodes to execution nodes. This migration is necessary to deploy the automation mesh.
```

File: downstream/modules/platform/proc-upgrade-installer.adoc [hunk 21]
```diff
@@ -64,6 +72,21 @@ node_type=execution
...
+----
+automationhub_api_token=<api_token>
+----
++
+* Generate a new API token, and invalidate any existing tokens, by setting the `generate_automationhub_token` flag to `True`:
++
+[source,options="nowrap",subs=attributes+]
+----
+generate_automationhub_token=True
+----
```
Comment (oraNod) [inline downstream/modules/platform/proc-upgrade-installer.adoc:89]: Does it need to be in caps?

Comment (TarikFD) [reply downstream/modules/platform/proc-upgrade-installer.adoc:89]: Just to keep it consistent with the rest of the doc, which all have it capitalized

Comment (oraNod) [reply downstream/modules/platform/proc-upgrade-installer.adoc:89]: Coolness. I was actually wondering if booleans were case sensitive but there's a pretty good explanation on stack overflow here: https://stackoverflow.com/questions/47877464/how-exactly-does-ansible-parse-boolean-variables

TL;DR I was just curious if there was a technical reason but `True` is good to go.

**Review by oraNod — APPROVED** [review]
LGTM. Nice work @TarikFD

File: downstream/modules/platform/con-aap-upgrade-planning.adoc [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+= {PlatformNameShort} upgrade planning
+
+[role="_abstract"]
+Before you begin the upgrade process, review the following considerations to plan and prepare your {PlatformNameShort} deployment:
+
+[discrete]
+== {ControllerNameStart}
+
+* Even if you have a valid license from a previous version, you must provide your credentials or a subscriptions manifest upon upgrading to the latest version of {ControllerName}.
+* If you need to upgrade Red Hat Enterprise Linux and {ControllerName}, you must first backup and restore of your {ControllerName} data.
```
Comment (lburcher7) [inline downstream/modules/platform/con-aap-upgrade-planning.adoc:14]: Unless I'm missing info, I think there is a typo in "restore of". It should be just "restore". Or should it be "you must first do a backup and restore of..."?
